src/langchain_google_alloydb_pg/model_manager.py

```diff
@@ -0,0 +1,180 @@
+from sqlalchemy.ext.asyncio import AsyncEngine
+
+
+class AlloyDBModelManager:
+    """Manage models to be used with google_ml_integration Extension.
+    Refer to [Model Endpoint Management](https://cloud.google.com/alloydb/docs/ai/model-endpoint-overview).
+    """
+
+    def __init__(
+        self,
+        engine: AsyncEngine,
+    ):
+        """AlloyDBModelManager constructor.
+        Args:
+            engine (AlloyDBEngine): Connection pool engine for managing connections to Postgres database.
+        """
+
+        self._engine = engine
+
+        self._validate()
+
+    async def alist_model(
+        self, model_id: str = "textembedding-gecko@001"
+    ) -> Sequence[RowMapping]:
+        """Lists the model details for a specific model_id.
+
+        Raises:
+            :class:`DBAPIError <sqlalchemy.exc.DBAPIError>`: if model has not been created.
+        """
+        result = await self._engine._run_as_async(self._alist_model(model_id=model_id))
+        return result
+
+    async def amodel_info_view(self) -> Sequence[RowMapping]:
+        """Lists all the models and its details."""
+        results = await self._engine._run_as_async(self._amodel_info_view())
+        return results
+
+    async def acreate_model(self, model_id: str, model_provider: str, **kwargs) -> None:
+        """Creates a custom text embedding model.
+
+        Raises:
+            :class:`DBAPIError <sqlalchemy.exc.DBAPIError>`: if argument names mismatch create_model function specification.
+        """
+        await self._engine._run_as_async(
+            self._acreate_model(model_id, model_provider, **kwargs)
+        )
+
+    async def adrop_model(self, model_id: str) -> None:
+        """Removes a text embedding model."""
+        await self._engine._run_as_async(self._adrop_model(model_id))
+
+    def _validate(self) -> None:
+        """Private function to validate prerequisites.
+
+        Raises:
+            Exception if google_ml_integration EXTENSION is not 1.3.
+            Exception if google_ml_integration.enable_model_support DB Flag not set.
+        """
+        extension_version = self._engine._run_as_sync(self._fetchGoogleMLExtension())
+        db_flag = self._engine._run_as_sync(self._fetchDBFlag())
+        if float(extension_version) < 1.3:
+            raise Exception("google_ml_integration EXTENSION is not 1.3")
+        if db_flag != "on":
+            raise Exception(
+                "google_ml_integration.enable_model_support DB Flag not set."
+            )
+
+    async def _query_db(self, query: str) -> Sequence[RowMapping]:
+        """Queries the Postgres database through the engine.
+
+        Raises:
+            Exception if the query is not a returning type."""
+        async with self._engine._pool.connect() as conn:
+            result = await conn.execute(text(query))
+            result_map = result.mappings()
+            results = result_map.fetchall()
+        return results
+
+    async def _alist_model(
```

Comment:
We should also make all mehtods that access the pool directly double underscore so users don't try to use them directly.